cpp/src/arrow/ipc/adapter.cc

```diff
@@ -309,66 +309,31 @@ class RecordBatchWriter : public ArrayVisitor {
     return Status::OK();
   }
 
-  Status Visit(const Int8Array& array) override {
-    return VisitFixedWidth<Int8Array>(array);
-  }
-
-  Status Visit(const Int16Array& array) override {
-    return VisitFixedWidth<Int16Array>(array);
-  }
-
-  Status Visit(const Int32Array& array) override {
-    return VisitFixedWidth<Int32Array>(array);
-  }
-
-  Status Visit(const Int64Array& array) override {
-    return VisitFixedWidth<Int64Array>(array);
-  }
-
-  Status Visit(const UInt8Array& array) override {
-    return VisitFixedWidth<UInt8Array>(array);
-  }
-
-  Status Visit(const UInt16Array& array) override {
-    return VisitFixedWidth<UInt16Array>(array);
-  }
-
-  Status Visit(const UInt32Array& array) override {
-    return VisitFixedWidth<UInt32Array>(array);
-  }
-
-  Status Visit(const UInt64Array& array) override {
-    return VisitFixedWidth<UInt64Array>(array);
-  }
-
-  Status Visit(const HalfFloatArray& array) override {
-    return VisitFixedWidth<HalfFloatArray>(array);
-  }
-
-  Status Visit(const FloatArray& array) override {
-    return VisitFixedWidth<FloatArray>(array);
-  }
-
-  Status Visit(const DoubleArray& array) override {
-    return VisitFixedWidth<DoubleArray>(array);
-  }
+#define VISIT_FIXED_WIDTH(TYPE) \
```

Comment:
Just a sidenote: When someone has a minute, we should think of how we could do this using `std::enable_if` instead of macros. I expect this to be much more compact.